File: automations/seo-optimizer/supabase/functions/seo-optimizer-analyst/categories/decay.ts
// Category A: Decay — YoY click drops on previously-performing URLs.

import type { ClientAnalysisContext, OpportunityCandidate } from "../types.ts";

const MIN_BASELINE_CLICKS = 50;
const MIN_DROP_PCT = 0.20;

interface UrlAgg {
  clicks: number;
  clicks_prev: number;
  impressions: number;
  impressions_prev: number;
  queries: Map<string, { clicks: number; clicks_prev: number; pos: number; pos_prev: number; n: number }>;
}

export function detect(ctx: ClientAnalysisContext): OpportunityCandidate[] {
  const byUrl = new Map<string, UrlAgg>();
  for (const r of ctx.gscRows) {
    if (!byUrl.has(r.url)) {
      byUrl.set(r.url, { clicks: 0, clicks_prev: 0, impressions: 0, impressions_prev: 0, queries: new Map() });
    }
    const slot = byUrl.get(r.url)!;
    slot.clicks += r.clicks;
    slot.impressions += r.impressions;
    slot.clicks_prev += r.clicks_prev ?? 0;
    slot.impressions_prev += r.impressions_prev ?? 0;
    if (!slot.queries.has(r.query)) {
      slot.queries.set(r.query, { clicks: 0, clicks_prev: 0, pos: 0, pos_prev: 0, n: 0 });
    }
    const q = slot.queries.get(r.query)!;
    q.clicks += r.clicks;
    q.clicks_prev += r.clicks_prev ?? 0;
    q.pos += r.position;
    q.pos_prev += r.position_prev ?? 0;
    q.n += 1;
  }

  const out: OpportunityCandidate[] = [];
  for (const [url, slot] of byUrl) {
    if (slot.clicks_prev < MIN_BASELINE_CLICKS) continue;
    const drop = slot.clicks_prev - slot.clicks;
    if (drop <= 0) continue;
    const dropPct = drop / slot.clicks_prev;
    if (dropPct < MIN_DROP_PCT) continue;

    const topLost = Array.from(slot.queries.entries())
      .filter(([, v]) => v.clicks_prev > v.clicks)
      .map(([q, v]) => ({
        query: q,
        clicks: v.clicks,
        clicks_prev: v.clicks_prev,
        clicks_delta: v.clicks - v.clicks_prev,
        position: Number((v.pos / Math.max(v.n, 1)).toFixed(2)),
        position_prev: Number((v.pos_prev / Math.max(v.n, 1)).toFixed(2)),
      }))
      .sort((a, b) => a.clicks_delta - b.clicks_delta)
      .slice(0, 5);

    const ci = ctx.contentItemByUrl.get(url);
    out.push({
      category: "decay",
      articleUrl: url,
      articleTitle: ci?.title ?? "",
      articleLanguage: ci?.language ?? null,
      contentItemId: ci?.contentItemId ?? null,
      evidence: {
        clicks_current: slot.clicks,
        clicks_yoy: slot.clicks_prev,
        drop_clicks: drop,
        drop_pct: Number(dropPct.toFixed(3)),
        top_queries_lost: topLost,
      },
      recommendationSummary:
        `Artículo en caída — perdió ${drop} clicks YoY (${(dropPct * 100).toFixed(0)}%). ` +
        `Refrescar secciones que cubrían: ${topLost.slice(0, 3).map(q => q.query).join(", ")}.`,
      recommendationDetails: {
        action: "refresh_content",
        focus_queries: topLost.map(q => q.query),
        rationale: "Queries con caída YoY indican que el contenido perdió relevancia o frescura.",
      },
      trafficPotential: drop * 1.5,
      confidence: "high",
      effortLevel: "medium",
      dedupeSignature: url,
    });
  }
  return out;
}

File: automations/seo-optimizer/02-agents/analyst/categories/decay.py
"""Category A: Decay — YoY click drops on previously-performing URLs.

Detect rule:
  - URL had >= 50 clicks YoY (baseline matters; don't waste effort on tiny URLs)
  - Current period clicks dropped >= 20% vs YoY
  - Compute top queries lost

Score = clicks_lost × 1.5 (high weight: proven URL needs rescue).
Confidence: high.  Effort: medium (refresh sections).
"""
from __future__ import annotations

from collections import defaultdict

from analyst.types import ClientAnalysisContext, OpportunityCandidate


MIN_BASELINE_CLICKS = 50
MIN_DROP_PCT = 0.20


def detect(ctx: ClientAnalysisContext) -> list[OpportunityCandidate]:
    # Aggregate per URL
    by_url: dict[str, dict] = defaultdict(lambda: {
        "clicks": 0, "clicks_prev": 0,
        "impressions": 0, "impressions_prev": 0,
        "queries": [],
    })

    for r in ctx.gsc_rows:
        slot = by_url[r.url]
        slot["clicks"] += r.clicks
        slot["impressions"] += r.impressions
        # If we have prev period data (set by gsc_ingestor in the row)
        # NOTE: GscRow doesn't carry _prev; we need to get them from the DB join.
        # In practice the analyst loads from gsc_url_query_metrics which has _prev
        # columns. For now we rely on a separate aggregation that the handler will
        # pass via ctx. See handler.run() — it builds gsc_rows from the DB rows
        # that already have clicks_prev populated.
        slot["clicks_prev"] += getattr(r, "clicks_prev", 0) or 0
        slot["impressions_prev"] += getattr(r, "impressions_prev", 0) or 0
        slot["queries"].append(r)

    out: list[OpportunityCandidate] = []
    for url, slot in by_url.items():
        clicks = slot["clicks"]
        clicks_prev = slot["clicks_prev"]

        if clicks_prev < MIN_BASELINE_CLICKS:
            continue
        drop = clicks_prev - clicks
        if drop <= 0:
            continue
        drop_pct = drop / clicks_prev
        if drop_pct < MIN_DROP_PCT:
            continue

        # Top queries lost (largest delta in clicks)
        per_query: dict[str, dict] = defaultdict(lambda: {
            "clicks": 0, "clicks_prev": 0,
            "position": 0.0, "position_prev": 0.0,
            "count": 0,
        })
        for r in slot["queries"]:
            q = per_query[r.query]
            q["clicks"] += r.clicks
            q["clicks_prev"] += getattr(r, "clicks_prev", 0) or 0
            q["position"] += r.position
            q["position_prev"] += getattr(r, "position_prev", 0) or 0
            q["count"] += 1

        top_lost = sorted(
            (
                {
                    "query": q,
                    "clicks": v["clicks"],
                    "clicks_prev": v["clicks_prev"],
                    "clicks_delta": v["clicks"] - v["clicks_prev"],
                    "position": round(v["position"] / max(v["count"], 1), 2),
                    "position_prev": round(v["position_prev"] / max(v["count"], 1), 2),
                }
                for q, v in per_query.items()
                if v["clicks_prev"] > v["clicks"]
            ),
            key=lambda x: x["clicks_delta"],
        )[:5]

        ci = ctx.content_item_by_url.get(url)
        out.append(OpportunityCandidate(
            category="decay",
            article_url=url,
            article_title=ci.title if ci else "",
            article_language=ci.language if ci else None,
            content_item_id=ci.content_item_id if ci else None,
            evidence={
                "clicks_current": clicks,
                "clicks_yoy": clicks_prev,
                "drop_clicks": drop,
                "drop_pct": round(drop_pct, 3),
                "top_queries_lost": top_lost,
            },
            recommendation_summary=(
                f"Artículo en caída — perdió {drop} clicks YoY ({drop_pct:.0%}). "
                f"Refrescar secciones que cubrían: {', '.join(q['query'] for q in top_lost[:3])}."
            ),
            recommendation_details={
                "action": "refresh_content",
                "focus_queries": [q["query"] for q in top_lost],
                "rationale": "Queries con caída de clicks YoY indican que el contenido perdió relevancia o frescura.",
            },
            traffic_potential=float(drop) * 1.5,
            confidence="high",
            effort_level="medium",
            dedupe_signature=url,
        ))

    return out
